Add LockSingleton for ECC lock

posterzh opened the pull request:
**Why**
To prevent ECC test fail, add the LockSingleton.  Use this singleton lock whenever the ECC is accessed.


**How**
Using Posix IPC, create a system wide semaphore to lock the ECC.
Also let hm-pyhelper provide convenient `@ecc_lock` decorator to keep the main logic simple and neat.

**References**
https://github.com/NebraLtd/hm-diag/issues/167

marvinmarnold commented:
@posterzh the demos are nice but we should have unit tests for all new code, especially things this complex.

I took the liberty of [adding two](cc64fff) (sorry for breaking flake8). `test_ecc_lock` is working correctly. But `test_lock_singleton_failure` is hanging for an unknown reason when running `pytest hm_pyhelper/tests/test_lock_singleton.py`. Can you help me debug? Can you then finish implementing the test (you should just be able to uncomment the rest of the code).?

posterzh commented:
@marvinmarnold The reason for the hangout is because there are no resources available. I updated the `LockSingleton` constructor to set the count of resources explicitly if it's set in the parameter.
And I added 2 test cases for `LockSingleton`, `test_lock_singleton_basic` and `test_lock_singleton_racing`. The result of the `test_lock_singleton_racing` looks like below.
![image](https://user-images.githubusercontent.com/91350131/139615474-d4ba7570-e725-4d43-b060-a70b9ca57c7b.png)

marvinmarnold commented on hm_pyhelper/lock_singleton.py:
```diff
@@ -11,17 +11,30 @@
+                                            initial_value=self._resource_count)
+            # Override the resource count if it's explicitly set. Otherwise the previous value
+            # of the resource count is preserved.
+            if resource_count and resource_count != self._sem.value:
```
This code smells funny to me. Did you find any documentation online that say this is a common pattern?

marvinmarnold replied:
What value do we even get in wrapping `posix_ipc.Semaphore`? Why doesn't ecc_lock just use `posix_ipc.Semaphore` directly?

posterzh replied:
Make sense. I refactored the this method to be easier to understand and added the comments how to use.

marvinmarnold commented on hm_pyhelper/lock_singleton.py:
```diff
@@ -11,17 +11,30 @@
 class LockSingleton(object):
```
It doesn't seem like `LockSingleton` is still a good name for this. It is now possible to create many different instances. I think `IpcLock` or `InterprocessLock` are more reflective of it now.

posterzh replied:
`InterprocessLock` sounds good. :+1:

marvinmarnold commented on hm_pyhelper/lock_singleton.py:
```diff
@@ -0,0 +1,106 @@
+def ecc_lock(func):
+    """Returns an ECC LOCK decorator.
+    """
+    lock = LockSingleton(LOCK_ECC)
```
This should have a timeout value of 1-2seconds.

posterzh replied:
I added `timeout` parameter.

posterzh commented on hm_pyhelper/tests/test_lock_singleton.py:
```diff
@@ -69,32 +72,34 @@ def fast_task():
+        Start a slow running thread and then try to acquire a shared
+        lock. Expect the acquire call to throw an exception.
+        """
         lock = LockSingleton('test')
```
@marvinmarnold I want to set the resource_count explicitly here. 
`lock = LockSingleton('test', resource_count=1)`

posterzh commented:
@marvinmarnold The available_resources of a `InterprocessLock` get reset on every restarting the system or docker container. It's tested in Ubuntu 20.04 desktop and diagnostics container in a Hotspot. I wrote about this in the Readme.md and comment section of `interprocess_lock.py`.

![image](https://user-images.githubusercontent.com/91350131/139713937-57d6b185-ac33-4b7b-85b1-fbe08892e1a5.png)

After restarting the `diagnostics` service.
![image](https://user-images.githubusercontent.com/91350131/139714398-46771843-cee9-47ee-ba65-cb8a273c5108.png)

vpetersson commented:
@posterzh What happens if the process crashes *after* the lock is set? Wouldn't we then get an infinite lock? I see there is a timeout for *acquiring* the lock, but not for resetting the lock.

robputt reviewed (approved):
Nice :-).  I think all of Marvin's requested changes have been completed. 👍

marvinmarnold reviewed (approved):
I'll implement the changes in hm-diag now.

BTW @posterzh did we ever actually test that this does in fact lock across two or more different containers on balena?
